### PR TITLE
Improve FileSystemError by adding associated path

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Mirrors https://github.com/apple/swift-tools-support-core/pull/167 change to the TSC package.

I've recently stumbled upon this error when running `swift build`:

```
error: noEntry
```

This is the exact and only output of this `swift build` invocation. Unfortunately, I think it's impossible to diagnose without an attached debugger. It's also hard to pinpoint where exactly the error comes from in the source code, since `FileSystemError.noEntry` is thrown from multiple places.

In my opinion, for an error value to be helpful it should have associated information attached to it. In this case, it would make sense for almost all cases of `FileSystemError` to have an associated `AbsolutePath` value.

`FileSystemError` now also has to be explicitly declared `Equatable` to make the tests compile, but previously it was `Equatable` anyway, although implicitly by virtue of being an enum with no associated values.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
https://github.com/apple/swift-tools-support-core/pull/167
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
apple/swift-tools-support-core#167
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
apple/swift-tools-support-core#167
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
apple/swift-tools-support-core#167
apple/sourcekit-lsp#348
@swift-ci please smoke test

File: swift-tools-support-core/Sources/TSCBasic/FileSystem.swift [hunk 1]
```diff
@@ -488,6 +504,9 @@ private class LocalFileSystem: FileSystem {
 //
 /// Concrete FileSystem implementation which simulates an empty disk.
 public class InMemoryFileSystem: FileSystem {
+    
+    /// Private internal representation of a file system node.
+    /// Not threadsafe.
```
Comment (abertelrud) [inline swift-tools-support-core/Sources/TSCBasic/FileSystem.swift:509]: These differences are all from syncing the vendored TSC, right?  For historical reasons, it would be clearer to make the vendoring sync be a separate commit, I think.

File: swift-tools-support-core/Sources/TSCBasic/WritableByteStream.swift [hunk 2]
```diff
@@ -672,16 +672,20 @@ public final class LocalFileOutputByteStream: FileOutputByteStream {
     /// The pointer to the file.
     let filePointer: UnsafeMutablePointer<FILE>
 
-    /// True if there were any IO error during writing.
-    private var error: Bool = false
+    /// Set to an error value if there were any IO error during writing.
+    private var error: FileSystemError?
```
Comment (abertelrud) [inline swift-tools-support-core/Sources/TSCBasic/WritableByteStream.swift:676]: This is such an improvement.  Thanks!

**Review by abertelrud — COMMENTED** [review]
In order to keep the history reasonably clear, and in case anything needs to be reverted, I suggest doing the TSC vendoring sync commit as a separate commit (can be in the same PR).

**Comment by abertelrud** [comment]
Also, I think the tests will keep failing until the TSC changes are merged, since at least some of the testing uses the actual TSC and not the vendored copy, I think.

File: swift-tools-support-core/Sources/TSCBasic/FileSystem.swift (hunk 1 above)
Comment (MaxDesiatov) [reply swift-tools-support-core/Sources/TSCBasic/FileSystem.swift:509]: The vendoring sync is already separate in 27bd6a8f7bf815ac3f8dfbe3242f352473930a60, or is there anything that I missed?

**Comment by neonichu** [comment]
Yes, self-hosted tests are using the main branch of TSC and do not support cross-repository testing. These jobs are optional, though, so not relevant for merging any PRs.

File: swift-tools-support-core/Sources/TSCBasic/FileSystem.swift (hunk 1 above)
Comment (MaxDesiatov) [reply swift-tools-support-core/Sources/TSCBasic/FileSystem.swift:509]: The subsequent SwiftPM update is in 68a729caa23d8b92585844def531ed98767bd5cc, and there's a small test fix separated in 6506c84ee38fd27a945df99376ca6f79a3f64dd0, but I can squash these last two if you'd like.

Comment (abertelrud) [reply swift-tools-support-core/Sources/TSCBasic/FileSystem.swift:509]: Ah, right:  it was a mixture of familiar and unfamiliar diffs, but that was only because the new diffs are actually in TSC and so are being vendored along with the previous diffs.  I see what my mistake was there.  Please ignore the comment, I just misunderstood (I think I also expected the commit to be something like "syncing TSC" rather than "Improve FileSystemError by adding associated path", which made me think that it was something more than vendoring.

Maybe just squashing the new diffs here (unless you want them in the history) so that this can be merged without squashing would be good.  Basically, at the end it would be good to end up with one diff in the final history for vendoring TSC and another for the SwiftPM diffs.  Sorry for the extra trouble.

**Comment by MaxDesiatov** [comment]
apple/swift-tools-support-core#167
apple/sourcekit-lsp#348
@swift-ci please smoke test

File: swift-tools-support-core/Sources/TSCBasic/FileSystem.swift (hunk 1 above)
Comment (MaxDesiatov) [reply swift-tools-support-core/Sources/TSCBasic/FileSystem.swift:509]: No problem, thanks for the clarification. I've squashed the last two commits and changed the title of the first one to mention the vendored TSC copy sync. Let me know if anything else is needed here.

**Comment by abertelrud** [comment]
Nice, thank you!  This looks good to me!

**Review by abertelrud — APPROVED** [review]
